File: authentication-python-demo/ocrcheck.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""易盾身份认证OCR信息查询接口python示例代码
接口文档: http://dun.163.com/api.html
python版本：python2.7
运行:
    1. 修改 SECRET_ID,SECRET_KEY,BUSINESS_ID 为对应申请到的值
    2. $ python ocr_check_api_demo.py
"""
__author__ = 'yidun-dev'
__version__ = '0.1-dev'

import hashlib
import time
import random
import urllib
import urllib2
import json

class TextCheckAPIDemo(object):
    """易盾身份认证OCR信息查询接口接口示例代码"""
    API_URL = "https://verify.dun.163yun.com/v1/ocr/check"
    VERSION = "v1"

    def __init__(self, secret_id, secret_key, business_id):
        """
        Args:
            secret_id (str) 产品密钥ID，产品标识
            secret_key (str) 产品私有密钥，服务端生成签名信息使用
            business_id (str) 业务ID，易盾根据产品业务特点分配
        """
        self.secret_id = secret_id
        self.secret_key = secret_key
        self.business_id = business_id

    def gen_signature(self, params=None):
        """生成签名信息
        Args:
            params (object) 请求参数
        Returns:
            参数签名md5值
        """
        buff = ""
        for k in sorted(params.keys()):
            buff += str(k)+ str(params[k])
        buff += self.secret_key
        return hashlib.md5(buff).hexdigest()

    def check(self, params):
        """请求易盾接口
        Args:
            params (object) 请求参数
        Returns:
            请求结果，json格式
        """
        params["secretId"] = self.secret_id
        params["businessId"] = self.business_id
        params["version"] = self.VERSION
        params["timestamp"] = int(time.time() * 1000)
        params["nonce"] = int(random.random()*100000000)
        params["signature"] = self.gen_signature(params)

        try:
            params = urllib.urlencode(params)
            request = urllib2.Request(self.API_URL, params)
            content = urllib2.urlopen(request, timeout=1).read()
            return json.loads(content)
        except Exception, ex:
            print "调用API接口失败:", str(ex)

if __name__ == "__main__":
    """示例代码入口"""
    SECRET_ID = "your_secret_id" # 产品密钥ID，产品标识
    SECRET_KEY = "your_secret_key" # 产品私有密钥，服务端生成签名信息使用，请严格保管，避免泄露
    BUSINESS_ID = "your_business_id" # 业务ID，易盾根据产品业务特点分配
    text_api = TextCheckAPIDemo(SECRET_ID, SECRET_KEY, BUSINESS_ID)

    params = {
        "picType": "1",
        "frontPicture": "http://www.xxx.com/xxx.jpg",
        "backPicture": "http://www.xxx.com/xxx.jpg",
    }
    ret = text_api.check(params)
    if ret is not None:
        if ret["code"] == 200:
            status = ret["result"]["status"]
            taskId = ret["result"]["taskId"]
            responsedetail = ret['result']['ocrResponseDetail']
            if status == 1:
                ocrName = ""
                ocrCardNo = ""
                expireDate = ""
                birthDay = ""
                nation = ""
                authority = ""
                gender = ""
                ocrAvatar = ""
                if responsedetail["ocrCardNo"] is not None:
                    ocrCardNo = responsedetail["ocrCardNo"].encode("utf-8")
                if responsedetail['ocrName'] is not None:
                    ocrName = responsedetail['ocrName'].encode("utf-8")
                if responsedetail['expireDate'] is not None:
                    expireDate = responsedetail['expireDate'].encode("utf-8")
                if responsedetail['birthDay'] is not None:
                    birthDay = responsedetail['birthDay'].encode("utf-8")
                if responsedetail['nation'] is not None:
                    nation = responsedetail['nation'].encode("utf-8")
                if responsedetail['gender'] is not None:
                    gender = responsedetail['gender'].encode("utf-8")
                if responsedetail['ocrAvatar'] is not None:
                    ocrAvatar = responsedetail['ocrAvatar'].encode("utf-8")
                if responsedetail['authority'] is not None:
                    authority = responsedetail['authority'].encode("utf-8")
                print "taskId=%s，查询成功,name = %s , cardNo = %s , expireDate = %s ,gender = %s ,nation = %s,birthDay = %s,ocrAvatar=%s,authority=%s" \
                    % (taskId.encode("utf-8"), ocrName, ocrCardNo, expireDate,gender, nation, birthDay,ocrAvatar,authority)
            elif status == 2:
                print "taskId=%s，解析失败,上传图片为非身份证图片或识别失败" % taskId.encode("utf-8")
        else:
            print "ERROR: ret.code=%s" %(ret['code'])

File: authentication-php-demo/ocrcheck.php
<?php
/** 产品密钥ID，产品标识 */
define("SECRETID", "your_secret_id");
/** 产品私有密钥，服务端生成签名信息使用，请严格保管，避免泄露 */
define("SECRETKEY", "your_secret_key");
/** 业务ID，易盾根据产品业务特点分配 */
define("BUSINESSID", "your_business_id");
/** 易盾身份认证服务身份证实名认证在线检测接口地址 */
define("API_URL", "https://verify.dun.163yun.com/v1/ocr/check");
/** api version */
define("VERSION", "v1");
/** API timeout*/
define("API_TIMEOUT", 2);
/** php内部使用的字符串编码 */
define("INTERNAL_STRING_CHARSET", "auto");
/**
 * 计算参数签名
 * $params 请求参数
 * $secretKey secretKey
 */
function gen_signature($secretKey, $params)
{
    ksort($params);
    $buff = "";
    foreach ($params as $key => $value) {
        if ($value !== null) {
            $buff .= $key;
            $buff .= $value;
        }
    }
    $buff .= $secretKey;
    return md5($buff);
}
/**
 * 将输入数据的编码统一转换成utf8
 * @params 输入的参数
 */
function toUtf8($params)
{
    $utf8s = array();
    foreach ($params as $key => $value) {
        $utf8s[$key] = is_string($value) ? mb_convert_encoding($value, "utf8", INTERNAL_STRING_CHARSET) : $value;
    }
    return $utf8s;
}
/**
 * 易盾身份认证服务OCR识别服务在线检测请求接口简单封装
 * $params 请求参数
 */
function check($params)
{
    $params["secretId"] = SECRETID;
    $params["businessId"] = BUSINESSID;
    $params["version"] = VERSION;
    $params["timestamp"] = sprintf("%d", round(microtime(true) * 1000));// time in milliseconds
    $params["nonce"] = sprintf("%d", rand()); // random int
    $params = toUtf8($params);
    $params["signature"] = gen_signature(SECRETKEY, $params);
    // var_dump($params);
    $options = array(
        'http' => array(
            'header' => "Content-type: application/x-www-form-urlencoded\r\n",
            'method' => 'POST',
            'timeout' => API_TIMEOUT, // read timeout in seconds
            'content' => http_build_query($params),
        ),
    );
    $context = stream_context_create($options);
    $result = file_get_contents(API_URL, false, $context);
    if ($result === FALSE) {
        return array("code" => 500, "msg" => "file_get_contents failed.");
    } else {
        return json_decode($result, true);
    }
}
// 简单测试
function main()
{
    echo "mb_internal_encoding=" . mb_internal_encoding() . "\n";
    $params = array(
        "picType" => "1",
        "frontPicture" => "http://xxx.com/xxx.jpg",
        "backPicture" => "http://xxx.com/xxx.jpg",
    );
    $ret = check($params);
    var_dump($ret);
    if ($ret["code"] == 200) {
        $status = $ret["result"]["status"];
        $taskId = $ret["result"]["taskId"];
        if ($status == 1) {
            $ocrName = "";
            $ocrCardNo = "";
            $expireDate = "";
            $gender = "";
            $nation = "";
            $birthday = "";
            $ocrAvatar = "";
            $authority = "";
            $ocrResponseDetail = $ret["result"]['ocrResponseDetail'];
            if ($ocrResponseDetail["ocrCardNo"]) {
                $ocrCardNo = $ocrResponseDetail["ocrCardNo"];
            }
            if ($ocrResponseDetail["ocrName"]) {
                $ocrName = $ocrResponseDetail['ocrName'];
            }
            if ($ocrResponseDetail["expireDate"]) {
                $expireDate = $ocrResponseDetail['expireDate'];
            }
            if ($ocrResponseDetail["gender"]) {
                $gender = $ocrResponseDetail['gender'];
            }
            if ($ocrResponseDetail["nation"]) {
                $nation = $ocrResponseDetail['nation'];
            }
            if ($ocrResponseDetail["birthday"]) {
                $birthday = $ocrResponseDetail['birthday'];
            }
            if ($ocrResponseDetail["ocrAvatar"]) {
                $ocrAvatar = $ocrResponseDetail['ocrAvatar'];
            }
            if ($ocrResponseDetail["authority"]) {
                $authority = $ocrResponseDetail['authority'];
            }
            echo "taskId={$taskId},查询成功,name = $ocrName , cardNo =  $ocrCardNo , expireDate = $expireDate ,gender = $gender ,nation = $nation,birthday = $birthday,ocrAvatar = $ocrAvatar,authority=$authority ";
        } else if ($status == 2) {
            echo "taskId={$taskId}，解析失败,上传图片为非身份证图片或识别失败";
        }
    } else {
        var_dump($ret); // error handler
    }
}
main();
?>